### ISSUE TITLE
Rework/simplify ListClipper drop impl

### ISSUE THREAD

**Issue by dbr** [imgui-rs/imgui-rs#610]
If you call `.end` on the list clipper, it'll crash when the destructor also calls the same

I don't think this happened in the last imgui-rs release, but happens in latest main branch (e.g 69fb34f0f8682ada766f7623c5ca626c219d9991)

The `long_list` example works fine as-is, but crashes if you do this:

```
$ git diff
diff --git a/imgui-examples/examples/long_list.rs b/imgui-examples/examples/long_list.rs
index 2204c53..fb7bab4 100644
--- a/imgui-examples/examples/long_list.rs
+++ b/imgui-examples/examples/long_list.rs
@@ -26,6 +26,7 @@ fn main() {
                         ui.text(&lots_of_words[row_num as usize]);
                     }
                 }
+                clipper.end();
             });
     });
 }
```

**Comment by dbr** [comment]
The crash happens in `bool ImGuiListClipper::Step()` on this line, because `data` is NULL:

```
if (data->StepNo == 0 && table != NULL && !table->IsUnfrozenRows)
```

The crashing call to step is happening via the Drop impl on `ListClipper`, which does this:

```rust
    fn drop(&mut self) {
        if !self.step() {
            unsafe {
                sys::ImGuiListClipper_destroy(self.list_clipper);
            };
        } else if !thread::panicking() {
            panic!(
```

This might be a regression in Dear ImGui, but..

Currently the `ListClipper` drop behaviour seems inconvinient - ignoring this bug, it'll panic if you don't exhaust the token, so I think something like

```
while clip.step() {
    ui.text(format!("Thing: {}", some_call()?));
}
```
..would panic in the case `some_call()?` returns early, which would be surprising at best

Given this..

1. isn't impacting any released imgui-rs version
2. the token is the last one with a panicing drop
3. the upstream API has changed
4. workaround in current `main` version is simple (just don't call `.end()`, as per long_list example)

Then it seems like we should just update the `ListClipper` to be inline with the permissive-dropping and new upstream API

**Comment by dbr** [comment]
This most likely relates to https://github.com/ocornut/imgui/issues/4822 in some way (although not exactly as that problem was hitting an intentional asser)

**Comment by dbr** [comment]
Worked around sufficiently in c50767f7e6e644db5f8fd4c66ba788d4d91e5b4c to close this - issue in imgui-rs repo will solve this properly

**Comment by ocornut** [comment]
Pushed a fix upstream for this and #611 
https://github.com/ocornut/imgui/commit/d9e60d29e1cab8c0b5370f92ce73b992ac77b3f1

**Comment by dbr** [comment]
> Then it seems like we should just update the ListClipper to be inline with the permissive-dropping and new upstream API

This indeed seems like the right thing to do as per https://github.com/imgui-rs/imgui-rs/issues/611#issuecomment-1044903235

The drop impl should just call end if it's not already - the extra call to check `Step() == false` is a bad idea (there's nothing wrong with stopping it half-way through a table, for example)

While changing this, should also investigate if the list clipper should act as an iterator, so instead of

```rust
let mut clip = imgui::ListClipper::new(table.total_rows() as i32).begin(ui);

while clip.step() {
    for row_num in clip.display_start()..clip.display_end() {
        ui.text("...")
    }
}
```
we can instead write:
```rust
let mut clip = imgui::ListClipper::new(table.total_rows() as i32).begin(ui);

for row in clip.iter_rows() {
    ui.text("...")
}
```

Should be easy to retain both interfaces as both are useful
